Feature: Collaborative Notes Pane for the Editor screen

EstoesMoises opened the pull request:
Hello everyone! I'm super happy to be writing this PR as it's one of my first open-source contributions :) **Plus,** this is a feature that I really want to have as part of DecapCMS.

Meet **the Notes Pane!**

---

### **Summary**

* Introduced a **Notes Pane** in the Editor (only for the Editorial Workflow), allowing users to add, edit, resolve, and delete notes linked to entries.
* Notes are managed per entry and support author info, timestamps, and resolution status.
* Notes are synchronized with backend storage and displayed in the dedicated UI pane (the Notes Pane).
* Added localization (English only) for all note-related actions and notifications.

Currently, the Notes feature has only been fully implemented on the **GitHub backend** and the **Test Backend**. I've also written some code for the Proxy Backend, but the feature will not work with `local_backend` as **it requires the Editorial Workflow.**

In essence, the goal of this feature works as a layer of abstraction for PR issue comments. With this we take more advantage of the information that is recorded in each PR issue. This aligns well with the philosophy behind the Editorial Workflow, particularly the (awesome) work Decap has done around PR and branch handling.

I made this change after following the discussion on [[Issue 52](https://github.com/decaporg/decap-cms/issues/52)](https://github.com/decaporg/decap-cms/issues/52), and personally Decap is a tool that I really want to use for some Documentation projects, the only thing I felt was missing was a "comments" feature, which is collaboration aspect that other popular platforms like Google Docs provide. But well, Decap is cooler ¯\_(ツ)_/¯ so it should have one as well.

---

### **Test Plan / Demonstration**

**Notes Pane UI within the Editor:**

<img width="2546" height="1262" alt="Screenshot 2025-07-28 214457" src="https://github.com/user-attachments/assets/9fde330b-dff5-4cd3-ad12-84cacaa4e6fb" />

**GitHub Comments:**

<img width="1673" height="1258" alt="Screenshot 2025-07-28 214711" src="https://github.com/user-attachments/assets/2601e309-7a8d-42e2-a1da-6b5c8ca417da" />

**Resolving Notes adds an HTML comment to the GitHub PR Issue Comment:**

<img width="1279" height="503" alt="Screenshot 2025-07-28 214723" src="https://github.com/user-attachments/assets/a00f5625-406d-400d-b656-b3d2b8c351fe" />

<img width="1141" height="353" alt="Screenshot 2025-07-28 214732" src="https://github.com/user-attachments/assets/29336386-81e7-48af-91d6-a746ce14b1eb" />

This passes all current tests, and I've modified some test files to account for the Notes actions. There’s still a pending change to add backend/implementation-specific testing.

---

### **Current state / room for improvement**

There is definitely room for improvement in this MVP. A few features that come to mind, and that are definitely on my roadmap (and I’d be more than happy to collaborate with others to make them happen) are:

* Polling new comments every few seconds/minutes
* As discussed in [[Issue 52](https://github.com/decaporg/decap-cms/issues/52)](https://github.com/decaporg/decap-cms/issues/52), this pane could be extended to show recent user activity, for example, when an entry has been reviewed, published, or unpublished.
* Implementation across all backends (currently only GitHub is supported as a production backend)
* Better unit testing

One of the main reasons I wanted to publish this in its current state is to get some early feedback before diving into more work. I’m really looking forward to hearing the maintainers’ thoughts and collaborating to make the Notes Pane feature as solid as possible.

---

### **Checklist**

Please add an `x` inside each checkbox:

* [x] I have read the [[contribution guidelines](https://github.com/decaporg/decap-cms/blob/main/CONTRIBUTING.md)](https://github.com/decaporg/decap-cms/blob/main/CONTRIBUTING.md).

---

### **A picture of a cute animal (not mandatory but encouraged)**

My dog, Luna :)

<img width="300" height="300" alt="image" src="https://github.com/user-attachments/assets/9129e1cd-d5cb-4e9a-9a4a-c04ef3cc4896" />

EstoesMoises committed:
feat: added interface for EditorNotesPane

EstoesMoises committed:
feat: crud with Redux complete for EditorNotesPane

EstoesMoises committed:
feat: notes functionality connected to Test Backend

EstoesMoises committed:
feat: notespane now immediately saves the note and it works with proxy backend

EstoesMoises committed:
feat: notespane now integrated with Github backend

EstoesMoises committed:
feat: adding notes to test backend

EstoesMoises committed:
chore: adding english locale to Note toasts

EstoesMoises committed:
chore: linting + notes will only work on EditorialWorkflow

EstoesMoises committed:
feat: notes can be unresolved

EstoesMoises committed:
fix: removing char limit on notes

EstoesMoises committed:
feat: avatar now visible in notes

EstoesMoises committed:
fix: now getting author and avatar immediately on note creation + notification added

EstoesMoises committed:
feat: you can only edit or delete your own notes

EstoesMoises committed:
chore: adding error handling for API calls and fixing colours in NoteForm

EstoesMoises committed:
fix: passing tests

EstoesMoises committed:
fix: automated test account for Notes

martinjagodic committed:
Merge branch 'main' into 52-notes-pane

EstoesMoises commented:
Hey @martinjagodic, I'm planning to tackle some of the action items that make sense from Copilot and to review this and ensure it is production-ready for the **Github Backend** 

Leaving the base for others to add compatibility with the other backends if they are interested.

The action items I'm thinking of before moving this from Draft PR to Ready are:
- Tackling copilot recommendations
- Add the new tests for my code
- Add a way to GET the notes when other users add them, probably using GitHub Webhooks if possible, and if not, a button to manually reload the notes.

I believe that with that done, I would consider it ready to be implemented in Main after your review.

Let me know your thoughts on this and if you'd like to add anything else, or if you have any feedback.

Thanks very much!

martinjagodic commented:
@EstoesMoises sounds good! How do you debug this while developing, as it doesn't work with the local server?

EstoesMoises commented:
> @EstoesMoises sounds good! How do you debug this while developing, as it doesn't work with the local server?

@martinjagodic I've been swapping backends by changing the config.yml from the dev-test folder - https://github.com/decaporg/decap-cms/blob/main/dev-test/config.yml

For debugging, I created and connected my GitHub account and I use a GitHub repository I made for this purpose. 

Please let me know if there is a different approach that you suggest.

Side note: So far, I've tested the Test Backend and GitHub Backend. When I started implementing this feature, I also considered creating functionality for the local backend by creating a .json for notes, but I reverted those changes to make this feature only available on Editorial Workflow.

martinjagodic commented:
I had a really quick check and these are my first impressions:

- Notes should be opt-in. I suggest `editor.notes: true/false` (false is default)
- Should work with i18n. The toggle button should be different, because now it's the same icon that toggles side-by-side locale editing.
- All entries should have notes, not only unpublished ones.

EstoesMoises commented:
> I had a really quick check and these are my first impressions:
> 
> * Notes should be opt-in. I suggest `editor.notes: true/false` (false is default)
> * Should work with i18n. The toggle button should be different, because now it's the same icon that toggles side-by-side locale editing.
> * All entries should have notes, not only unpublished ones.

Perfect! - Thanks for the feedback Martin. I will work on this

EstoesMoises commented:
Just wanted to leave a quick note - due to some unexpected life events, it might take me up to two weeks to get back with an update. I’ll still be working on the changes as soon as I can and I'm looking forward and aiming to get this feature completed 👍

EstoesMoises commented:
Hello everyone, due to unexpected circumstances, it has been a while since my last update, but I'm still here and will get this updated very soon 👍  I'm just sending this message to confirm that this is not abandoned, and I'm very excited about getting this feature ready :P

EstoesMoises committed:
fix: inverted logic for initializing notesVisible

EstoesMoises committed:
fix: more robust handleBlur function for notes

EstoesMoises committed:
feat: notes pane feature is now opt-in depending on config

EstoesMoises committed:
chore: changed icon for notesPane to quote

EstoesMoises committed:
feat: notes now use Github issues  making notes usable for both published and unpublished entries

EstoesMoises committed:
fix: removed condition that would loadNotes() twice

EstoesMoises committed:
feat: lifecycle for github issues when publishing/unpublishing entries

EstoesMoises committed:
chore: format

EstoesMoises committed:
chore: added tests for notes on editorialWorkflow specs

EstoesMoises committed:
chore: added tests for Notes Github implementation

EstoesMoises committed:
chore: updated config specification test to account for the editor.notes option

EstoesMoises commented:
Hey @martinjagodic this PR is ready for your review.

- I've added the required tests for the Github implementation.
- Changed from using PR Comments to create Github issues (and issues comments), this way I could enable Notes to both Published and Unpublished entries.
- Github issues are closed when entries are published and are reopened if that same entry is unpublished.

Let me know your thoughts on this ! Looking forward to your feedback

EstoesMoises committed:
Merge branch 'main' into 52-notes-pane

martinjagodic commented:
Those are good points. Maybe someone else shares a fresh opinion, but you convinced me. 

Do you think having an option in config to choose between issues or PRs would be too complicated to implement?

maybe:
```yml
editor:
  notes: 
    enable: true
    source: issue # or pull_request
```

An additional option is to have both the above config and `notes: true/false`, and it would just take issues as default.

We could also have `notes: issue/pull_request/false`, but this is mixing enums and booleans, which is not ideal.

EstoesMoises committed:
fix: github issues are only created after the first note is added

EstoesMoises committed:
fix: notes now show dates in absolute time

EstoesMoises committed:
style: increased padding of NoteList, NoteItem and AddNoteForm

EstoesMoises committed:
feat: github notes now have a convenient link to the source (github issues)

EstoesMoises committed:
feat: decapNotes now have a polling system to retrieve new notes automatically

EstoesMoises committed:
fix: notes polling system now works well with first-ever Note

EstoesMoises committed:
chore: adding more context to console.log of decapNotes functionality when no notes exist yet

EstoesMoises committed:
chore: formatting

EstoesMoises committed:
chore: fixed with correct data structure in DecapNotes test

EstoesMoises committed:
chore: added tests for decapNotes polling system

EstoesMoises committed:
Merge branch 'main' into 52-notes-pane

martinjagodic commented:
Awesome! The polling system is a great solution. I will take some time this week to test this properly.

martinjagodic reviewed (changes requested):
Sorry it took me so long.

I get some 401 errors when testing on a public GitHub repo [poslovnimedij/decap-cms-tester](https://github.com/poslovnimediji/decap-cms-tester/tree/preview-pane). On the preview-pane branch, there is a minimal reproduction config. It comes with a build of your branch.

I can see the notes pane, but there is a 401 error on the first load

```
Request URL: https://preview-pane--decap-cms-tester.netlify.app/.netlify/git/github/users/demshy
Request Method: GET
Status Code: 401 Unauthorized
```

The request is about a user who is not me, therefore the problem.

When I try to add a note, I get this toast error: "ui.toast.onFailToAddNote". No network requests are triggered.

As you can see, I could not test much. The toggling between preview and notes is ok. I think notes need a different icon than the quote, though.  A document or a document with a pencil is a better option.

EstoesMoises commented:
Heyy @yanthomasdev it's been difficult on my end because of lack of time. But I would like to continue working on this in order to complete the contribution. I intend to pick things up next weekend.

yanthomasdev commented:
That would be great @EstoesMoises, no rush! 💜

EstoesMoises committed:
Merge branch 'main' into 52-notes-pane

EstoesMoises committed:
fix: changed Notes icon

EstoesMoises committed:
chore: add error for Note adding failure

EstoesMoises commented:
Hey @martinjagodic, it has been a while. I hope you are doing very well.

I have changed the icon for the notes pane. Let me know if that works for you.

On a different note, I tried to replicate what you mentioned and followed the instructions to use https://github.com/poslovnimediji/decap-cms-tester/tree/preview-pane. The notes functionality I built does not work with a local_backend (it requires the editorial workflow), so I am not completely sure how to test it using that tool.

The notes pane does work for me when running `npm run start` on this very branch and connecting it to a GitHub repository. The feature I built currently supports only GitHub as a backend and does not support Git Gateway.

I'll provide the dev-test/config.yml I use below.

I'd like to fix the issue you are experiencing tho, could you help me with instructions on how to properly test this?

Thanks in advance

CC: @yanthomasdev 

```
backend:
  name: github
  branch: main
  repo: estoesmoises/test

publish_mode: editorial_workflow
media_folder: static/media
public_folder: /media
editor:
  notes: true
collections:
  - name: posts
    label: Posts
    label_singular: 'Post'
    folder: content/posts
    create: true
    slug: '{{year}}-{{month}}-{{day}}-{{slug}}'
    fields:
      - label: Template
        name: template
        widget: hidden
        default: post
      - label: Title
        name: title
        widget: string
      - label: 'Cover Image'
        name: 'image'
        widget: 'image'
        required: false
      - label: Publish Date
        name: date
        widget: datetime
        format: 'YYYY-MM-DDTHH:mm'
        default: 1970-01-01T01:00
      - label: Description
        name: description
        widget: text
      - label: Category
        name: category
        widget: string
      - label: Body
        name: body
        widget: markdown
      - label: Tags
        name: tags
        widget: list
  - name: pages
    label: Pages
    label_singular: 'Page'
    folder: content/pages
    create: true
    slug: '{{slug}}'
    fields:
      - label: Template
        name: template
        widget: hidden
        default: page
      - label: Title
        name: title
        widget: string
      - label: Draft
        name: draft
        widget: boolean
        default: true
      - label: Body
        name: body
        widget: markdown

```

EstoesMoises committed:
Merge branch 'main' into 52-notes-pane

EstoesMoises committed:
fix: note.author defaults to empty string

EstoesMoises committed:
Merge branch 'main' into 52-notes-pane

martinjagodic commented:
@EstoesMoises I'm so sorry to keep you waiting :( We were merging some old open things now, so I didn't have time to check this.

EstoesMoises committed:
Merge branch 'main' into 52-notes-pane

yanthomasdev committed:
Merge branch 'main' into 52-notes-pane

yanthomasdev committed:
fix: replace uuid with crypto.randomUUID()

yanthomasdev committed:
fix: add missing await

yanthomasdev committed:
fix: gate notes loading by effective editor config

yanthomasdev committed:
fix: stop notes polling on editor unmount

yanthomasdev commented:
Thank you @EstoesMoises, great work! I've made a few changes to your branch to fix some stuff I found, could you take a look? I believe it should soon be good to go.

EstoesMoises commented:
Looks good to me ! 

And that's great news :D @yanthomasdev
